numpy/typing/__init__.py

```diff
@@ -64,6 +64,31 @@
 ndarray
 ~~~~~~~
 
+The ``ndarray`` type is generic with respect to both its shape and data type.
+The shape parameter is a tuple of ints or subtypes of int.
+
+The data type field is a subclass of ``np.dtype``, usually via parametrization
+with a scalar type: e.g.
+``Float1D = np.ndarray[tuple[int], np.dtype[np.float64]]``.
+
+Array creation cannot yet detect the shape parameter from arguments; it must be
+declared, e.g.:
+
+.. code-block:: python
+
+    > from typing import Literal, NewType
+    > Width = NewType("Width", int)
+    > Height = NewType("Height", int)
+    >
+    > GrayImg = np.ndarray[tuple[Width, Height], np.dtype[np.uint8]]
+    > RGBImg = np.ndarray[tuple[Width, Height, Literal[3]], np.dtype[np.uint8]]
+    >
+    > def colorize(img: GrayImg) -> RGBImg: ...
+    > 
+    > x: GrayImg = np.array([[1, 2, 3], [4, 5, 6]])
+    > assert_type(colorize(x), RGBImg)  # type check passes
+
+
```

Comment:
This looks very similar to the text change notes, so my comments there also apply here.

Reply:
Following those comments from above, if I remove the detail from the change notes, are you ok with a more detailed explanation here?

Reply:
I don't think this is needed at all actually, as this is directed to the few of us that work on numpy typing.
If we're unsure of the specifics of the shape-type-parameter, then we can just look at the `TypeVar` itself (at least, that's what I'd do; but that's probably because my attention span is in units of attoseconds).

Reply:
That all makes sense, but until a week ago, I wasn't really planning on working more deeply with numpy typing.  I was exactly the user you described above: one who sees this change as a feature that allows me to shape-type my own functions, even if I have to `cast` the output.

I'm ok removing the docs for now, though.  I agree it could be confusing for users who, as soon as they do something like `arr + 1`, lose the shape type information.